src/mptool.f90

```diff
@@ -143,5 +290,343 @@ pure function cross_product(a,b)
+      call mpi_file_set_view(fh, offset, real_type, real_type, 'native', mpi_info_null, ierr)
+      call mpi_file_write(fh, data_5, local_size, real_type, status, ierr)
+
+    endif
+
+    if(present(data_6)) then
+
+      ! Barrier to synchronize processes
+      call mpi_barrier(mpi_comm_world, ierr)
+
+      offset = offset + total_size
+
+      ! write the data_6 to the file
+      call mpi_file_set_view(fh, offset, real_type, real_type, 'native', mpi_info_null, ierr)
+      call mpi_file_write(fh, data_6, local_size, real_type, status, ierr)
+
+    endif
+
+    ! close the file
+    call mpi_file_close(fh, ierr)
+
+    if(nrank==0) print*,'<< ',filename
+
+  end subroutine pwrite_2darray
+
+  subroutine pread_1darray(filename,data2read)
+
+    ! arguments
+    character(len=*),intent(in) :: filename
+    real(mytype),intent(inout),allocatable :: data2read(:)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_1darray
+
+
+  subroutine pread_2darray(filename,data2read)
+
+    ! arguments
+    character(len=*),intent(in) :: filename
+    real(mytype),intent(inout),allocatable :: data2read(:,:)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    local_size=size(data2read,2)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_2darray
+
+  ! this function return the number add from all processors before it
```

Comment:
This is looking like a duplicate of `psum`

Reply:
I think there is no need for two functions doing the same job

Reply:
done with the function.

Reply:
Please look at the input and output of the function `psum_integer` and `prelay`. Please describe the difference between both. It seems they are doing providing the exact same output.

Reply:
Ok